### PR TITLE
Using new ImGui's IO event API (1.87) through IMGUI_DISABLE_OBSOLETE_KEYIO

### PR DESCRIPTION
**Pull request by johannesugb**
With `#define IMGUI_DISABLE_OBSOLETE_KEYIO` its new event API should be used internally, so that all the components and examles are prepared for the future change (if not already changed ImGui-internally implementation-wise). 

Any reservations against `#define IMGUI_DISABLE_OBSOLETE_KEYIO` or shall we leave it like that?

`imfilebrowser.h` had to be updated to support  `#define IMGUI_DISABLE_OBSOLETE_KEYIO`, which leads to the following variables not being declared in `imgui.h`:
```
#ifndef IMGUI_DISABLE_OBSOLETE_KEYIO
    int         KeyMap[ImGuiKey_COUNT];             // [LEGACY] Input: map of indices into the KeysDown[512] entries array which represent your "native" keyboard state. The first 512 are now unused and should be kept zero. Legacy backend will write into KeyMap[] using ImGuiKey_ indices which are always >512.
    bool        KeysDown[512];                      // [LEGACY] Input: Keyboard keys that are pressed (ideally left in the "native" order your engine has access to keyboard keys, so you can use your own defines/enums for keys).
#endif
```

### CODE REVIEW

**Comment by ifl0w** [comment]
Why not add `IMGUI_DISABLE_OBSOLETE_FUNCTIONS` as well? You will have to rename `ImGuiKey_KeyPadEnter` to `ImGuiKey_KeypadEnter`.

Is it really necessary to add it to all those CPP files? I thought adding it to `imgui_manager.hpp` was enough since that is the main interface to ImGui?

**Comment by mrumpelnik** [comment]
I think @ifl0w has a point here. Framework users include `gvk.hpp`, which already defines `IMGUI_DISABLE_OBSOLETE_KEYIO`, therefore having it in every `cpp` file is kind of redundant

According to the transition guide, ImGui will also define `IMGUI_DISABLE_OBSOLETE_FUNCTIONS` with `IMGUI_DISABLE_OBSOLETE_KEYIO`. So better also prepare the `imgui_manager` code for this

**Comment by johannesugb** [comment]
Okay, the reason for `#define IMGUI_DISABLE_OBSOLETE_KEYIO` everywhere is the following:

Initially, I planned to have `<imgui.h>` included totally optionally. I.e., if an application does not use ImGui, `<imgui.h>` would not be included at all. This was initially possible because the **header** file of `imgui_manager` did not require any code from ImGui at all. If the implementation of `imgui_manager` was not used by an application, no ImGui code was compiled at all, because the **implemenation** of `imgui_manager` was not used.

That has changed not too long ago due to the usage of `ImTextureID` in `imgui_manager.hpp`.

There are two routes ahead:
1. Leave it ike proposed in this PR + the proposed code changes from @ifl0w and @mrumpelnik => We'd have the ImGui-related `#define`s only in `imgui_manager.hpp` in this case.
2. Forward-declare `ImTextureID` in `imgui_manager.hpp` and retain the possibility of not compiling any ImGui code. Every example would have to make the ImGui-`#define`s on its own in this case.

@ifl0w and @mrumpelnik what are your opinions on this? Could the second approach be advantageous (option to not compile ImGui code, maybe lower compile times) or shall we just go the more straight-forward way of the first approach?

**Comment by mrumpelnik** [comment]
Ok, makes sense. I like the idea to retain the possibility of not compiling any ImGui code but having `IMGUI_DISABLE_OBSOLETE_KEYIO` and `IMGUI_DISABLE_OBSOLETE_FUNCTIONS` in every example and also framework user code? seems odd. Especially since these defines will become basically "useless" in the future when `imgui.h` defines them, which it will according to the transition guide.

How about
- Forward-declare `ImTextureID` in `imgui_manager.hpp` to retain the possiblity of not compiling ImGui
- Define `IMGUI_DISABLE_OBSOLETE_KEYIO` and `IMGUI_DISABLE_OBSOLETE_FUNCTIONS` in `imgui_manager.hpp`
- Include `gvk.hpp` before `imgui.h` in every example (I think this is already the case)

That should work fine and every framework user compiles the same internal ImGui code without modifying anything and in the future when `imgui.h` defines `IMGUI_DISABLE_OBSOLETE_KEYIO` and `IMGUI_DISABLE_OBSOLETE_FUNCTIONS` you can just delete it in `imgui_manger.hpp` and don't have to update every example again.
Of course, there is the problem with the include order and users getting it wrong but you kind of also have the same thing if users forget to define `IMGUI_DISABLE_OBSOLETE_KEYIO` and `IMGUI_DISABLE_OBSOLETE_FUNCTIONS` in there code right?

**Comment by ifl0w** [comment]
Fundamentally, I like the approach of @mrumpelnik since it combines the best of both worlds. And would go with that for now if you do not see a noteworthy issue with it.

A different solution I can think of is to define the flags in `CMakeLists.txt` with a short comment. IMO, this would be the cleanest solution as it does not pollute the code at all and is globally defined regarding of include order (i.e., I think the configuration should be done in the build system). I don't know if there is an equivalent solution for Visual Studio though (I guess there is).

**Comment by johannesugb** [comment]
Awesome, thanks for this nice discussion. 
If you agree with the changes pushed just a minute ago, I'll go ahead and merge into `master`.
@ifl0w did I put the defines at the right place within `CMakeLists.txt`?

File: CMakeLists.txt [hunk 1]
```diff
@@ -79,6 +79,10 @@ if (MSVC)
         ${gvk_IncludeScope} "/MP")
 endif(MSVC)
 
+# Disable obsolete functionality in ImGui:
+# (Can be removed once ImGui has deprecated these for good, and acts as if these were defined.)
+add_definitions(-DIMGUI_DISABLE_OBSOLETE_KEYIO -DIMGUI_DISABLE_OBSOLETE_FUNCTIONS)
```
Comment (ifl0w) [inline CMakeLists.txt:84]: add_definitions is superseded by newer functions that are more explicit and don't require the "-D".
I'd suggest replacing with the following: 
```suggestion 
target_compile_definitions(${PROJECT_NAME} PUBLIC IMGUI_DISABLE_OBSOLETE_KEYIO IMGUI_DISABLE_OBSOLETE_FUNCTIONS)
```

For more information see [CMake Docs](https://cmake.org/cmake/help/latest/command/target_compile_definitions.html?highlight=target_compile_definitions)

**Pull request merged**